src/Algebra/Graph/Undirected/UndirectedGraph.hs

```diff
@@ -0,0 +1,935 @@
+{-# LANGUAGE DeriveGeneric, RankNTypes #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.Undirected.UndirectedGraph
+-- Copyright  : (c) Andrey Mokhov 2016-2019
+-- License    : MIT (see the file LICENSE)
+-- Maintainer : andrey.mokhov@gmail.com
+-- Stability  : experimental
+--
+-- __Alga__ is a library for algebraic construction and manipulation of graphs
+-- in Haskell. See <https://github.com/snowleopard/alga-paper this paper> for the
+-- motivation behind the library, the underlying theory, and implementation details.
+--
+-- This module defines and abstraction over the core data type 'Graph' and associated algorithms.
+-- Undirected graphs satisfy all laws of the
+-- 'Algebra.Graph.Class.Undirected' type class, including the commutativity
+-- of 'connect'.
+-- For graphs that are known to be /non-empty/ at compile time, see
+-- "Algebra.Graph.NonEmpty". 'Graph' is an instance of type classes defined in
+-- modules "Algebra.Graph.Class" and "Algebra.Graph.HigherKinded.Class", which
+-- can be used for polymorphic graph construction and manipulation.
+--
+-----------------------------------------------------------------------------
+module Algebra.Graph.Undirected.UndirectedGraph (
+    -- * Algebraic data type for graphs
+    UndirectedGraph (..),
```

Comment:
Let's call this data type simply `Graph`. Users will use qualified modules to disambiguate.

Reply:
Also, let's not export the constructor and see how an abstract implementation will work.

Reply:
okay!